### PR TITLE
feat: [CDS-67674]: Multi Select in Allowed Values type runtime input and allow comma in the allowed values strings.

### PR DESCRIPTION
**Pull request by shalini-agr**
## Describe your changes

**Description:** Allow ',' in a string in allowed values and let the user select multiple values from allowed values in runtime input.

**Changes:** In the validator, use a regex to split allowed values and input fields such that it splits strings with a comma but not when a comma is within single or double quotes. For eg: The string `abc,\"blr,india\",\'ny,us\'` will be considered as the list of `["abc","blr,india","ny,us"]`

**Testing:** Tested these cases locally:

- Select multiple values in the run form (should be allowed to execute if each of them is in the allowed values list and fails otherwise)
- Comma within quotes of the string is considered a part of the string.
- The field with the above value is being used in a step during execution (same behavior is shown as above) (Different validator runs in this case)

Error message when input values are invalid:
Screenshot 2023-05-03 at 12 45 13 PM

## Checklist
- [x] I've documented the changes in the PR description.
- [x] I've tested this change either in PR or local environment.
- [x] If hashcheck failed I followed [the checklist](https://harness.atlassian.net/wiki/spaces/DEL/pages/21016838831/PR+Codebasehash+Check+merge+checklist) and updated this PR description with my answers to make sure I'm not introducing any breaking changes.

## Comment Triggers
<details>
  <summary>Build triggers</summary>

- Feature build: `trigger feature-build`
- 
  Specific builds
  - `trigger manager`
  - `trigger dms`
  - `trigger ng_manager`
  - `trigger cvng `
  - `trigger cmdlib`
  - `trigger template_svc`
  - `trigger events_fmwrk_monitor`
  - `trigger event_server`
  - `trigger change_data_capture`
  -  Trigger multiple builds together: For eg: `trigger dms, manager`
- Immutable delegate `trigger publish-delegate`
</details>

<details>
  <summary>PR Check triggers</summary>

You can run multiple PR check triggers by comma separating them in a single comment. e.g. `trigger ti0, ti1`

- Compile: `trigger compile`
- CodeformatCheckstyle: `trigger checkstylecodeformat`
  - CodeFormat: `trigger codeformat`
  - Checkstyle: `trigger checkstyle`
- MessageMetadata: `trigger messagecheck`
- File-Permission-Check: `trigger checkpermission`
- Recency: `trigger recency`
- BuildNumberMetadata: `trigger buildnum`
- Trigger CommonChecks: `trigger commonchecks`
- PMD: `trigger pmd`
- Copyright Check: `trigger copyrightcheck`
- Feature Name Check: `trigger featurenamecheck`
- UnitTests-ALL: `trigger utAll`
- UnitTests-0: `trigger ut0`
- UnitTests-1: `trigger ut1`
- UnitTests-2: `trigger ut2`
- UnitTests-3: `trigger ut3`
- UnitTests-4: `trigger ut4`
- UnitTests-5: `trigger ut5`
- UnitTests-6: `trigger ut6`
- UnitTests-7: `trigger ut7`
- UnitTests-8: `trigger ut8`
- UnitTests-9: `trigger ut9`
- CodeBaseHash: `trigger codebasehash`
- CodeFormatCheckstyle: `trigger checkstylecodeformat`
- SonarScan: `trigger ss`
- GitLeaks: `trigger gitleaks`
- Trigger all Checks: `trigger smartchecks`
- Go Build: `trigger gobuild`
- Validate_Reviews: `trigger review`
</details>

## PR check failures and solutions
https://harness.atlassian.net/wiki/spaces/BT/pages/21106884744/PR+Checks+-+Failures+and+Solutions

## [Contributor license agreement](https://github.com/harness/harness-core/blob/develop/CONTRIBUTOR_LICENSE_AGREEMENT.md)

- - -
This change is [Reviewable](https://reviewable.io/reviews/harness/harness-core/47083)

### CODE REVIEW

File: 953-yaml-commons/src/main/java/io/harness/pms/yaml/validation/RuntimeInputValuesValidator.java [hunk 1]
```diff
@@ -83,20 +83,30 @@ public String validateStaticValues(Object templateObject, Object inputSetObject)
...
         } else if (inputSetValidator.getValidatorType() == ALLOWED_VALUES) {
-          String[] allowedValues = inputSetValidator.getParameters().split(", *");
-          boolean matches = false;
-          for (String allowedValue : allowedValues) {
-            if (NGExpressionUtils.isRuntimeOrExpressionField(allowedValue)) {
-              return error;
-            } else if (allowedValue.equals(inputSetFieldValue)) {
-              matches = true;
+          List<String> allowedValues = AllowedValuesSplitter.split(inputSetValidator.getParameters());
+          List<String> inputFields = AllowedValuesSplitter.split(inputSetFieldValue);
```
Comment (archit-harness) [inline 953-yaml-commons/src/main/java/io/harness/pms/yaml/validation/RuntimeInputValuesValidator.java:87]: inputSetfieldValue if given as multi select is coming correctly now?

Comment (shalini-agr) [reply 953-yaml-commons/src/main/java/io/harness/pms/yaml/validation/RuntimeInputValuesValidator.java:87]: yes

**Comment by shalini-agr** [comment]
trigger commonchecks

**Review by archit-harness — CHANGES_REQUESTED** [review]
> Reviewed 6 of 7 files at r5, 2 of 2 files at r6, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/47083)* status: 8 file(s) to review. 1 change requests. 0 of 2 approvals obtained. 

___
*[`953-yaml-commons/src/main/java/io/harness/pms/yaml/validation/AllowedValuesSplitter.java` line 27 at r6](https://reviewable.io/reviews/harness/harness-core/47083#-NUVFPmHA1RGozGiDdmW:-NUVFPmHA1RGozGiDdmX:biwrjgj) ([raw file](https://github.com/harness/harness-core/blob/d769b908f99d89aa286223391a348c1d8e24f413/953-yaml-commons/src/main/java/io/harness/pms/yaml/validation/AllowedValuesSplitter.java#L27)):*
> ```Java
> 
>     for (int currentPosition = 0; currentPosition < input.length(); currentPosition++) {
>       if (currentPosition + 1 < input.length() && input.charAt(currentPosition) == '\\'
> ```

Add comments on each if
___
*[`953-yaml-commons/src/main/java/io/harness/pms/yaml/validation/AllowedValuesSplitter.java` line 40 at r6](https://reviewable.io/reviews/harness/harness-core/47083#-NUVFmZqCF2VfiNnd9j-:-NUVFmZqCF2VfiNnd9j0:b8clnmf) ([raw file](https://github.com/harness/harness-core/blob/d769b908f99d89aa286223391a348c1d8e24f413/953-yaml-commons/src/main/java/io/harness/pms/yaml/validation/AllowedValuesSplitter.java#L40)):*
> ```Java
> 
>     String lastValue = input.substring(startPosition);
>     if (!lastValue.equals(",")) {
> ```

this if condition not required



_Code quote:_
```Java
if (!lastValue.equals(",")) {
```
___
*[`953-yaml-commons/src/main/java/io/harness/pms/yaml/validation/RuntimeInputValuesValidator.java` line 87 at r3](https://reviewable.io/reviews/harness/harness-core/47083#gh-1180501843:-NUVEb5zEdN6n6T7jIjH:bb7f2mf) ([raw file](https://github.com/harness/harness-core/blob/5199f243d0ec9e14f8d09749eba58ef5e0332449/953-yaml-commons/src/main/java/io/harness/pms/yaml/validation/RuntimeInputValuesValidator.java#L87)):*
<details><summary><i>Previously, shalini-agr (Shalini Agrawal) wrote…</i></summary><blockquote>

yes
</blockquote></details>

done
___
*[`953-yaml-commons/src/main/java/io/harness/pms/yaml/validation/RuntimeInputValuesValidator.java` line 95 at r6](https://reviewable.io/reviews/harness/harness-core/47083#-NUVGzkZ-U9J4EQIHKhp:-NUVGzkZ-U9J4EQIHKhq:b8xh1sq) ([raw file](https://github.com/harness/harness-core/blob/d769b908f99d89aa286223391a348c1d8e24f413/953-yaml-commons/src/main/java/io/harness/pms/yaml/validation/RuntimeInputValuesValidator.java#L95)):*
> ```Java
>             for (String allowedValue : allowedValues) {
>               if (NGExpressionUtils.isRuntimeOrExpressionField(allowedValue)) {
>                 return error;
> ```

Reuse as a util method in 2 classes
___
*[`956-feature-flag-beans/src/main/java/io/harness/beans/FeatureName.java` line 746 at r6](https://reviewable.io/reviews/harness/harness-core/47083#-NUVF-Tf9b0aw0p4iLLf:-NUVF-Tf9b0aw0p4iLLg:b-xsu73e) ([raw file](https://github.com/harness/harness-core/blob/d769b908f99d89aa286223391a348c1d8e24f413/956-feature-flag-beans/src/main/java/io/harness/beans/FeatureName.java#L746)):*
> ```Java
>       "Will fire the artifact and manifest triggers for all the versions in the polling response instead of the latest only",
>       HarnessTeam.SPG),
>   SPG_MULTISELECT_AND_COMMA_IN_ALLOWED_VALUES(
> ```

PIE team

**Review by shalini-agr — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/47083)* status: 8 file(s) to review. 1 change requests. 4. unresolved discussions. 0 of 2 approvals obtained. 

___
*[`953-yaml-commons/src/main/java/io/harness/pms/yaml/validation/RuntimeInputValuesValidator.java` line 95 at r6](https://reviewable.io/reviews/harness/harness-core/47083#-NUVGzkZ-U9J4EQIHKhp:-NUVPJRNCrEdcFZI9bFD:b-pavjau) ([raw file](https://github.com/harness/harness-core/blob/d769b908f99d89aa286223391a348c1d8e24f413/953-yaml-commons/src/main/java/io/harness/pms/yaml/validation/RuntimeInputValuesValidator.java#L95)):*
<details><summary><i>Previously, archit-harness (Archit Singla) wrote…</i></summary><blockquote>

Reuse as a util method in 2 classes
</blockquote></details>

as discussed offline, cannot reuse has RuntimeInputValidator has some custom logic different from AllowedValuesValidator
___
*[`956-feature-flag-beans/src/main/java/io/harness/beans/FeatureName.java` line 746 at r6](https://reviewable.io/reviews/harness/harness-core/47083#-NUVF-Tf9b0aw0p4iLLf:-NUVN_c08qDyJAhi5BBS:b-896fix) ([raw file](https://github.com/harness/harness-core/blob/d769b908f99d89aa286223391a348c1d8e24f413/956-feature-flag-beans/src/main/java/io/harness/beans/FeatureName.java#L746)):*
<details><summary><i>Previously, archit-harness (Archit Singla) wrote…</i></summary><blockquote>

PIE team
</blockquote></details>

Done.
___
*[`953-yaml-commons/src/main/java/io/harness/pms/yaml/validation/AllowedValuesSplitter.java` line 27 at r6](https://reviewable.io/reviews/harness/harness-core/47083#-NUVFPmHA1RGozGiDdmW:-NUVN-vg-xOrV-o0GQc2:b-896fix) ([raw file](https://github.com/harness/harness-core/blob/d769b908f99d89aa286223391a348c1d8e24f413/953-yaml-commons/src/main/java/io/harness/pms/yaml/validation/AllowedValuesSplitter.java#L27)):*
<details><summary><i>Previously, archit-harness (Archit Singla) wrote…</i></summary><blockquote>

Add comments on each if
</blockquote></details>

Done.
___
*[`953-yaml-commons/src/main/java/io/harness/pms/yaml/validation/AllowedValuesSplitter.java` line 40 at r6](https://reviewable.io/reviews/harness/harness-core/47083#-NUVFmZqCF2VfiNnd9j-:-NUVN0QDBNPkgjmUs062:b-896fix) ([raw file](https://github.com/harness/harness-core/blob/d769b908f99d89aa286223391a348c1d8e24f413/953-yaml-commons/src/main/java/io/harness/pms/yaml/validation/AllowedValuesSplitter.java#L40)):*
<details><summary><i>Previously, archit-harness (Archit Singla) wrote…</i></summary><blockquote>

this if condition not required
</blockquote></details>

Done.

**Comment by shalini-agr** [comment]
trigger utAll

**Review by archit-harness — APPROVED** [review]
> Reviewed 5 of 5 files at r7, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/47083)* status: 8 file(s) to review. 1 of 2 approvals obtained.

**Review by brijeshdhakar — APPROVED** [review]
> Reviewed 3 of 7 files at r5, 5 of 5 files at r7, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/47083)* status: :shipit: complete! ✓

**Comment by shalini-agr** [comment]
trigger review

**Comment by shalini-agr** [comment]
trigger template_svc

**Pull request merged**